src/core/layout/qgslayoutitemchart.cpp

```diff
@@ -0,0 +1,390 @@
+{
+  if ( mSortExpression == expression )
+  {
+    return;
+  }
+
+  mSortExpression = expression;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::setSeriesList( const QList<QgsLayoutItemChart::SeriesDetails> &seriesList )
+{
+  if ( mSeriesList == seriesList )
+  {
+    return;
+  }
+
+  mSeriesList = seriesList;
+  refresh();
+
+  emit changed();
+}
+
+void QgsLayoutItemChart::draw( QgsLayoutItemRenderContext & )
+{
+}
+
+void QgsLayoutItemChart::paint( QPainter *painter, const QStyleOptionGraphicsItem *itemStyle, QWidget * )
+{
+  if ( !mLayout || !painter || !painter->device() )
+  {
+    return;
+  }
+
+  if ( !shouldDrawItem() )
+  {
+    return;
+  }
+
+  if ( !mPlot )
+    return;
+
+  QPaintDevice *paintDevice = painter->device();
+  if ( !paintDevice )
+    return;
+
+  QRectF thisPaintRect = rect();
+  if ( qgsDoubleNear( thisPaintRect.width(), 0.0 ) || qgsDoubleNear( thisPaintRect.height(), 0 ) )
+    return;
+
+  if ( mLayout->renderContext().isPreviewRender() )
+  {
+    if ( mNeedsGathering || mIsGathering )
+    {
+      if ( mNeedsGathering )
+      {
+        mNeedsGathering = false;
+        refreshData();
+      }
+
+      QgsScopedQPainterState painterState( painter );
+      painter->setClipRect( thisPaintRect );
+
+      painter->setBrush( QBrush( QColor( 125, 125, 125, 125 ) ) );
+      painter->drawRect( thisPaintRect );
+      painter->setBrush( Qt::NoBrush );
+      QFont messageFont;
+      messageFont.setPointSize( 12 );
+      painter->setFont( messageFont );
+      painter->setPen( QColor( 255, 255, 255, 255 ) );
+      painter->drawText( thisPaintRect, Qt::AlignCenter | Qt::AlignHCenter, tr( "Rendering chart" ) );
+      return;
+    }
+  }
+  else
+  {
+    if ( mNeedsGathering )
+    {
+      mNeedsGathering = false;
+      prepareGatherer();
+      QgsApplication::instance()->taskManager()->addTask( mGatherer.data() );
+      mGatherer->waitForFinished( 60000 );
+    }
+  }
+
+  const double scaleFactor = QgsLayoutUtils::scaleFactorFromItemStyle( itemStyle, painter );
+  const QSizeF size = mLayout->convertToLayoutUnits( sizeWithUnits() ) * scaleFactor;
+  if ( size.width() == 0 || size.height() == 0 )
+    return;
+
+  mPlot->setSize( size );
+
+  QgsScopedQPainterState painterState( painter );
+  painter->scale( 1 / scaleFactor, 1 / scaleFactor );
+
+  QgsRenderContext renderContext = QgsLayoutUtils::createRenderContextForLayout( mLayout, painter );
+  renderContext.setScaleFactor( scaleFactor );
+  renderContext.setExpressionContext( createExpressionContext() );
+
+  QgsPlotRenderContext plotRenderContext;
+  mPlot->render( renderContext, plotRenderContext, mPlotData );
+}
+
+void QgsLayoutItemChart::refresh()
+{
+  QgsLayoutItem::refresh();
+  mNeedsGathering = true;
+}
+
+void QgsLayoutItemChart::refreshData()
+{
+  mGathererTimer.start();
+}
+
+void QgsLayoutItemChart::gatherData()
+{
+  prepareGatherer();
+  QgsApplication::instance()->taskManager()->addTask( mGatherer.data() );
```

Comment:
there's cases in prepareGatherer where mGatherer may be null:

```suggestion
  if ( mGatherer )
  {
    QgsApplication::instance()->taskManager()->addTask( mGatherer.data() );
  }
```